adapters/beachfront/beachfront.go

```diff
@@ -197,182 +188,278 @@ func (a *BeachfrontAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *a
+		} else {
+			bannerReq.Domain = request.Site.Domain
+		}
 
-		appid = ext.AppId
-	} else if fmt.Sprintf("%s", reflect.TypeOf(ext.AppIds)) == "openrtb_ext.ExtImpBeachfrontAppIds" {
-		if media == openrtb_ext.BidTypeVideo && ext.AppIds.Video != "" {
-			appid = ext.AppIds.Video
-		} else if media == openrtb_ext.BidTypeBanner && ext.AppIds.Banner != "" {
-			appid = ext.AppIds.Banner
+		bannerReq.IsMobile = 0
+	}
+
+	if request.User != nil && request.User.ID != "" {
+		if bannerReq.User.ID == "" {
+			bannerReq.User.ID = request.User.ID
+		}
+	}
+
+	if request.User != nil && request.User.BuyerUID != "" {
+		if bannerReq.User.BuyerUID == "" {
+			bannerReq.User.BuyerUID = request.User.BuyerUID
 		}
-	} else {
-		error = errors.New("unable to determine the appId(s) from the supplied extension")
 	}
 
-	return appid, error
+	bannerReq.RequestID = request.ID
+	bannerReq.AdapterName = beachfrontAdapterName
+	bannerReq.AdapterVersion = beachfrontAdapterVersion
+
+	return bannerReq, errs
 }
 
-/*
-getBannerRequest, singular. A "Slot" is an "imp," and each Slot can have an AppId, so just one
-request to the beachfront banner endpoint gets all banner Imps.
-*/
-func getBannerRequest(request *openrtb.BidRequest) (beachfrontBannerRequest, []error) {
-	var bfr beachfrontBannerRequest
-	var errs = make([]error, 0, len(request.Imp))
+func impsToSlots(imps []openrtb.Imp, errs []error) (bannerRequest, int8, []error) {
+	var bfr bannerRequest
+	var secure int8 = 0
 
-	for i := 0; i < len(request.Imp); i++ {
+	for i := 0; i < len(imps); i++ {
+		var imp = imps[i]
+		if imp.Banner == nil {
+			continue
+		}
 
-		beachfrontExt, err := getBeachfrontExtension(request.Imp[i])
+		if imp.Secure != nil {
+			secure = *imp.Secure
+		} else {
+			secure = 0
+		}
+		beachfrontExt, err := getBeachfrontExtension(imp)
 
 		if err != nil {
-			errs = append(errs, err)
+			errs = append(errs, errors.New(fmt.Sprintf("%s (on banner imp id: %s)", err, imp.ID)))
 			continue
 		}
 
 		appid, err := getAppId(beachfrontExt, openrtb_ext.BidTypeBanner)
 
 		if err != nil {
-			// Failed to get an appid, so this request is junk.
-			errs = append(errs, err)
+			errs = append(errs, errors.New(fmt.Sprintf("%s (on banner imp id: %s)", err, imp.ID)))
 			continue
 		}
 
-		slot := beachfrontSlot{
+		slot := slot{
 			Id:       appid,
-			Slot:     request.Imp[i].ID,
+			Slot:     imp.ID,
 			Bidfloor: beachfrontExt.BidFloor,
 		}
 
 		if beachfrontExt.BidFloor <= minBidFloor {
 			slot.Bidfloor = 0
 		}
 
-		for j := 0; j < len(request.Imp[i].Banner.Format); j++ {
+		for j := 0; j < len(imp.Banner.Format); j++ {
+			slot.Sizes = append(slot.Sizes, size{
+				H: imp.Banner.Format[j].H,
+				W: imp.Banner.Format[j].W,
+			})
+		}
 
-			slot.Sizes = append(slot.Sizes, beachfrontSize{
-				H: request.Imp[i].Banner.Format[j].H,
-				W: request.Imp[i].Banner.Format[j].W,
+		if len(slot.Sizes) == 0 {
+			slot.Sizes = append(slot.Sizes, size{
+				W: *imp.Banner.W,
```

Comment:
What happens here if `imp.Banner.W` is nil?

Reply:
len(slot.Sizes)==0 would only be true at this point (L399) if imp.Banner.Format had been length 0. If I understand correctly, imp.Banner.Format can't be nil as defined in mxmcherry banner.go, so I'm not checking that. If both imp.Banner.Format were length 0, and therefore len(slot.Sizes)==0 is true, and imp.Banner.W or imp.Banner.H were nil, it would not have passed the check at /endpoints/openrtb2/auction.go L441 (line number may have changed of course, but in the function validateBanner), and the entire request would have been discarded with an error.

If I'm not mistaken.

Reply:
Makes sense. I was trying to figure out what could be causing your race condition, and this appeared a little suspicious. I thought we might have validation for that, but I wanted to at least flag this before diving deeper into your code. I believe I found a real issue below.

Reply:
I made a quick test case for that, and I do seem to be mistaken. It threw a segmentation fault.

Reply:
@hhhjort - So do you think that I should keep my check for defined H and W, which appears to only come into play with the test specifically made for it, but not in regular execution? Or delete both the check and the test?

Reply:
I suppose I slightly lean to deleting them, as it has been shown we have verification elsewhere. I don't have a strong feeling either way.